### ISSUE TITLE
glimmer samples Meta-Sample code_text broken styled text

### ISSUE THREAD

**Issue by MingScott** [AndyObtiva/glimmer-dsl-swt#25]
**Desktop (please complete the following information and make sure it matches [Glimmer Pre-Requisites](https://github.com/AndyObtiva/glimmer-dsl-swt#pre-requisites) before reporting an issue):**
 - CPU Architecture [e.g. x86_64 or AArch64]: x86_64
 - OS [e.g. Linux Ubuntu 20.04.1 LTS]:  Ubuntu 20.04.4 LTS
 - Java Version [e.g. JDK 16.0.2]: openjdk 17.0.4
 - JRuby Version [e.g. jruby-9.3.1.0]: jruby 9.3.6.0
 - Glimmer Libraries/Versions [e.g. glimmer 2.4.1 / glimmer-dsl-swt 4.21.2.0]:  glimmer-dsl-swt v4.24.1.3

**Describe the issue**
A clear and concise description of what the issue is, including expected behavior and actual result.

When loading the Meta-Sample application, the styled code text and line numbers are expected to be comprehensible text

Instead, they are boxes filled with unicode codepoints

**Comment by MingScott** [comment]
![glimmer_glitch](https://user-images.githubusercontent.com/29578947/184759464-6b5fa495-ee89-4486-a1b5-576386e3e6eb.png)
Here's what it looks like

**Comment by MingScott** [comment]
Additional context: It does appear to be code_text that's broken, as the hello_code_text sample is doing this:
![glimmer_code_text](https://user-images.githubusercontent.com/29578947/184771052-10df95c5-abfd-4029-899d-b5918082a053.png)

**Comment by AndyObtiva** [comment]
I am asking because `code_text` is based on `styled_text`, and there is a Hello sample for each of them. I'm glad you shared a screenshot of Hello, Code Text! with me, but I'd like to also see if the problem happens with Hello, Styled Text! , which uses a raw `styled_text` instead of using it indirectly through `code_text`.

**Comment by AndyObtiva** [comment]
By the way, the problem is most likely related to SWT and not Glimmer DSL for SWT. I think it has to do with compatibility with your Ubuntu 20.04.4 LTS. You may be missing the typically built-in GTK 3 library somehow (that happens if you customize your Linux to remove it).

**Comment by AndyObtiva** [comment]
One way of confirming if the problem is directly related to SWT running on your system is to build an SWT app in Java, using the `StyledText` widget, and then see if it displays unicode characters.

**Comment by MingScott** [comment]
Here's ``styled_text``, which appears to work... perfectly! The plot thickens

![styled_text](https://user-images.githubusercontent.com/29578947/184789216-4facd7dc-3181-4bc8-9f40-9bb33ffe13bc.png)

**Comment by MingScott** [comment]
I don't have this machine set up for java development at the moment, but considering that ``styled_text`` works, I don't see why the pure SWT version wouldn't

**Comment by AndyObtiva** [comment]
Thank you for sharing a screenshot of Hello, Styled Text!

The plot thickens indeed.

I guess now we've located the issue to be in Glimmer DSL for SWT, not SWT, so you don't need to do Java testing anymore.

The issue caused by the font used by `code_text`. It assumes you either have 'Consolas' or 'Courier'. I did not expect that some operating systems might be missing 'Courier', but it is obvious your platform is missing it. 

The issue can be fixed in one of two ways:
1. Install 'Courier' font to your Linux OS
2. Manually set another available font (try modified Hello, Code Text! example code below (including `font name: 'Helvetica'` inside `code_text` declarations) to prove this works by pasting it in the Glimmer Meta-Sample editor or running standalone, but change 'Helvetica' to a font that you know exists on your machine, preferably a monospace font)
3. Update Glimmer DSL for SWT to seek the next available monospace font (similar to 'Consolas' and 'Courier') when neither 'Consolas' nor 'Courier' are available.

Please try #2 to confirm that it works on your machine and share a screenshot with me. Ensure you replace 'Helvetica' with another font name, preferably a monospace font.

In the meantime, I will implement #3 as a good long-term fix to ensure `code_text` works on Linux installations that do not have the 'Courier' font. Once I have done so, I will close this issue.

**Modified Hello, Code Text! (including `font name: 'Helvetica'` inside `code_text` declarations)**

```ruby
require 'glimmer-dsl-swt'

class HelloCodeText
  include Glimmer::UI::CustomShell
  
  attr_accessor :ruby_code, :js_code, :html_code
  
  before_body do
    self.ruby_code = <<~RUBY
      greeting = 'Hello, World!'
      
      include Glimmer
      
      shell {
        text 'Glimmer'
        
        label {
          text greeting
          font height: 30, style: :bold
        }
      }.open
    RUBY
    
    self.js_code = <<~JS
      function greet(greeting) {
        alert(greeting);
      }
      
      var greetingString = 'Hello, World!';
      
      greet(greetingString);
      
      var moreGreetings = ['Howdy!', 'Aloha!', 'Hey!']
      
      for(var greeting of moreGreetings) {
        greet(greeting)
      }
    JS
    
    self.html_code = <<~HTML
      <html>
        <body>
          <section class="accordion">
            <form method="post" id="name">
              <label for="name">
                Name
              </label>
              <input name="name" type="text" />
              <input type="submit" />
            </form>
          </section>
        </body>
      </html>
    HTML
  end
  
  body {
    shell {
      minimum_size 640, 480
      text 'Hello, Code Text!'
      
      tab_folder {
        tab_item {
          fill_layout
          
          text 'Ruby (glimmer theme)'
          
          code_text(language: 'ruby', theme: 'glimmer', lines: true) {
            font name: 'Helvetica'
            text <=> [self, :ruby_code]
          }
        }
        
        tab_item {
          fill_layout
          
          text 'JavaScript (pastie theme)'
          
          code_text(:multi, :h_scroll, :v_scroll, language: 'javascript', theme: 'pastie', lines: {width: 2}) {
            font name: 'Helvetica'
            root {
              grid_layout(2, false) {
                margin_width 2
              }
              
              background :white
            }
            
            line_numbers {
              background :white
            }
            
            text <=> [self, :js_code]
          }
        }
        
        tab_item {
          fill_layout
          
          text 'HTML (github theme)'
          
          code_text(language: 'html', theme: 'github') { # default is lines: false
            font name: 'Helvetica'
            text <=> [self, :html_code]
          }
        }
      }
    }
  }
  
end

HelloCodeText.launch
```

**Comment by MingScott** [comment]
![hello_code_text_fixed](https://user-images.githubusercontent.com/29578947/184984827-489bd5fb-f944-42df-b0cd-66a8a8fc7934.png)
![fixed_meta_sample](https://user-images.githubusercontent.com/29578947/184985418-28a77689-982d-479f-b1c6-3b7f8e2daa9e.png)
![fixed_code](https://user-images.githubusercontent.com/29578947/184985650-a60ff8d4-fca3-4c2f-82df-d28ac1fce2c6.png)

Set both to explicitly call Ubuntu Mono, which seems to have fixed it

**Comment by AndyObtiva** [comment]
OK, good news!

The newly released [Glimmer DSL for SWT v4.24.2.2](https://rubygems.org/gems/glimmer-dsl-swt/versions/4.24.2.2) not only fixes the font issue of `code_text` on Linux, but it also now supports Dark Mode in `code_text` using the `glimmer_dark` theme (always forced when the OS is in Dark Mode). Alternatively, you can build your own Rouge theme just like [glimmer_dark](https://github.com/AndyObtiva/glimmer-dsl-swt/blob/master/lib/ext/rouge/themes/glimmer_dark.rb) and use it yourself, but it has to have the word “dark” in its name in order for `code_text` not to force `glimmer_dark` in Dark Mode.

![Screen Shot 2022-08-21 at 2 51 21 PM](https://user-images.githubusercontent.com/23052/185811866-a14fc7eb-8d59-4f2c-bbb4-adec7c83034c.png)
![Screen Shot 2022-08-21 at 2 51 22 PM](https://user-images.githubusercontent.com/23052/185811870-0a89231e-98cc-4c7b-9ee8-5e174dea457f.png)
![Screen Shot 2022-08-21 at 2 51 24 PM](https://user-images.githubusercontent.com/23052/185811871-e2c577c3-446e-4b91-a790-522ba6c91a25.png)
![Screen Shot 2022-08-21 at 2 51 33 PM](https://user-images.githubusercontent.com/23052/185811872-fab9f252-65e8-4b2c-bbc9-8c1fd3c7e10b.png)

Please test and confirm that the latest release works for you. Feel free to share screenshots if you like.

I am closing this issue for now unless you encounter further issues that you report, then we can reopen (or open other issues).

Cheers!
